src/H5FDonion_index.c

```diff
@@ -0,0 +1,935 @@
+ *              Failure: Zero (0) -- entry not found.
+ *                       Entry out pointer-pointer is unmodified.
+ *-----------------------------------------------------------------------------
+ */
+int
+H5FD__onion_revision_index_find(const H5FD_onion_revision_index_t *rix, uint64_t logical_page,
+                                const H5FD_onion_index_entry_t **entry_out)
+{
+    uint64_t key       = 0;
+    int      ret_value = 0;
+
+    FUNC_ENTER_PACKAGE_NOERR;
+
+    HDassert(rix);
+    HDassert(H5FD__ONION_REVISION_INDEX_VERSION_CURR == rix->version);
+    HDassert(rix->_hash_table);
+    HDassert(entry_out);
+
+    key = logical_page & (rix->_hash_table_size - 1);
+    HDassert(key < rix->_hash_table_size);
+
+    if (rix->_hash_table[key] != NULL) {
+        H5FD_onion_revision_index_hash_chain_node_t *node = NULL;
+
+        for (node = rix->_hash_table[key]; node != NULL; node = node->next) {
+            if (logical_page == node->entry_data.logical_page) {
+                *entry_out = &node->entry_data;
+                ret_value  = 1;
+                break;
+            }
+        }
+    }
+
+    FUNC_LEAVE_NOAPI(ret_value);
+} /* end H5FD__onion_revision_index_find() */
+
+/*-----------------------------------------------------------------------------
+ * Function:    H5FD__onion_revision_record_decode
+ *
+ * Purpose:     Attempt to read a buffer and store it as a revision record
+ *              structure.
+ *
+ *              Implementation must correspond with
+ *              H5FD__onion_revision_record_encode().
+ *
+ *              MUST BE CALLED TWICE:
+ *              On the first call, n_entries and comment_size in the
+ *              destination structure must all all be zero, and their
+ *              respective variable-length components (index_entry_list,
+ *              comment) must all be NULL.
+ *
+ *              If the buffer is well-formed, the destination structure is
+ *              tentatively populated with fixed-size values, and the number of
+ *              bytes read are returned.
+ *
+ *              Prior to the second call, the user must allocate space for the
+ *              variable-length components, in accordance with the associated
+ *              indicators (array of index-entry structures for
+ *              index_entry_list, of size n_entries; character arrays for
+ *              comment, allocated with the *_size number of bytes -- space
+ *              for NULL-terminator is included in _size).
+ *
+ *              Then the decode operation is called a second time, and all
+ *              components will be populated (and again number of bytes read is
+ *              returned).
+ *
+ * Return:      Success:    Number of bytes read from buffer
+ *              Failure:    0
+ *-----------------------------------------------------------------------------
+ */
+size_t
+H5FD__onion_revision_record_decode(unsigned char *buf, H5FD_onion_revision_record_t *record)
```

Comment:
Comment for future reference:

The decode and encode routines in this file should probably be aware of how large buf is so we don't run off the end of it on accident.